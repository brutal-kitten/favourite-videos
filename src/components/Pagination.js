import React, { Component } from 'react';
import ListOfVideos from './ListOfVideos';
import PaginationBox from './PaginationBox';
import PropTypes from 'prop-types';

class Pagination extends Component {

  static propTypes = {
    recalculatePages: PropTypes.bool.isRequired,
    recalculatePagesSetFalse: PropTypes.func.isRequired,
    showDemo: PropTypes.func.isRequired,
    deleteList: PropTypes.func.isRequired,
    sort: PropTypes.func.isRequired,
    listOfVideo: PropTypes.array.isRequired,
    addToFavorite: PropTypes.func.isRequired,
    removeFromFavorite: PropTypes.func.isRequired,
    deleteVideo: PropTypes.func.isRequired
  }

  state = {
    showFav: false,
    startIndex: 0,
    elementsPerPage: 10,
    totalPages: 1,
    currentPage: 1
  }

  //reculculate and set new state - favorites always starts from 1st page
  showFavorite = () => {

    let total = this.calculateTotalPagesNumber(
      this.props.listOfVideo.filter(item => (item.favorite === true)).length, this.state.elementsPerPage);
      this.setState({showFav: true, startIndex: 0, currentPage: 1, totalPages: total});
  }

  //change state so user will see all list instead of favorite
  returnToList = () => {

    this.setState({showFav: false})
  }

  //calls function that show demolist and change state so user will see full list
  beforeShowDemo = () => {

    this.props.showDemo();
    this.setState({showFav: false});
  }

  //calculate length of list and  with given number of items per page recalculate number of total pages and set new state
  setIndexes = (number) => {

    let length = this.filterList(this.props.listOfVideo).length;
    let totalPagesNumber = this.calculateTotalPagesNumber(length, number);
    this.setState({
      elementsPerPage: number,
      startIndex: 0,
      totalPages: totalPagesNumber,
      currentPage: 1});
  }

  //with given new current page reculculate start index and set new state
  setStartIndex = (currentPage) => {

    if (currentPage === 1 ) {
      this.setState({startIndex : 0, currentPage: currentPage });
    } else {
      let currentMinusOne = currentPage - 1;
      let index = parseInt(this.state.elementsPerPage, 10)*(currentMinusOne);
      this.setState({startIndex : index, currentPage: currentPage});
    }
  }

  //with given length of array and number of items per page calculate and return total number of page
  calculateTotalPagesNumber = (length, number) => {

    return (length % number === 0) ? (length / number) : (parseInt(length / number) + 1);
  }

  //create array of items for showing on curren page
  createList = (list) => {

    let filteredList = this.filterList(list);
    let start = parseInt(this.state.startIndex, 10);
    let number = parseInt(this.state.elementsPerPage, 10);
    let finish =  start + number;
    let slicedList = filteredList.slice(start, finish);
    return slicedList;
  }

  //return full array or only favorites - depends on state
  filterList = (list) => {

    let filteredList = (this.state.showFav ? (list.filter(item => (item.favorite === true))) : list);
    return filteredList;
  }

  checkForUpdate () {
    let length = this.filterList(this.props.listOfVideo).length;
    let start = parseInt(this.state.startIndex, 10);
    let elPerPage =  parseInt(this.state.elementsPerPage, 10);
    let totalPagesNumber = this.calculateTotalPagesNumber(length, elPerPage);

    //when user delete all items in array
    if (length === 0) {
      this.setState({totalPages: 1, currentPage: 1, startIndex: 0, elementsPerPage: 10});
      this.props.recalculatePagesSetFalse();
      } else if (start >= length) {
       // when user delete item and it was last item on the carrent page
       // and now user shoud see previous page
          let startIndex = elPerPage*(totalPagesNumber - 1);
          this.setState({totalPages: totalPagesNumber, startIndex: startIndex, currentPage: totalPagesNumber});
          this.props.recalculatePagesSetFalse();
        } else if (totalPagesNumber > this.state.totalPages) {
            //when user add item or items
            this.setState({totalPages: totalPagesNumber});
            this.props.recalculatePagesSetFalse();
        }
  }

  /* each time user delete item or add new item or delete full list recalculate
   and set information about current page, total pages number, startIndex */
  componentDidUpdate () {

    if (this.props.recalculatePages === true) {
      this.checkForUpdate();
    };
  }

  render () {

    return(
      <div className="page">
      <PaginationBox
        setIndexes={this.setIndexes}
        totalPages={this.state.totalPages}
        setStartIndex={this.setStartIndex}
        elementsPerPage={this.state.elementsPerPage}
        currentPage={this.state.currentPage}
      />
      <ListOfVideos
        showDemo={this.beforeShowDemo}
        deleteList={this.props.deleteList}
        showFavorite={this.showFavorite}
        showFav={this.state.showFav}
        returnToList={this.returnToList}
        sort={this.props.sort}
        list={this.createList(this.props.listOfVideo)}
        addToFavorite={this.props.addToFavorite}
        removeFromFavorite={this.props.removeFromFavorite}
        deleteVideo={this.props.deleteVideo}
        playVideo={this.props.playVideo}
        changeSearchResultError={this.props.changeSearchResultError}
        number={this.state.number}
      />
      </div>
    )
  }
}

export default Pagination
