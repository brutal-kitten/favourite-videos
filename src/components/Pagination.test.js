import React from 'react';
import ReactDOM from 'react-dom';
import Pagination from './Pagination.js';
import TestRenderer from 'react-test-renderer';
import { shallow, mount, render } from 'enzyme';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

const mockRecalculatePagesSetFalse = jest.fn();
const mockShowDemo = jest.fn();
const mockDeleteList = jest.fn();
const mockSort = jest.fn();
const mockAddToFavorite = jest.fn();
const mockRemoveFromFavorite = jest.fn();
const mockDeleteVideo = jest.fn();


const props = {
  recalculatePages: false,
  recalculatePagesSetFalse: mockRecalculatePagesSetFalse,
  showDemo: mockShowDemo,
  deleteList: mockDeleteList,
  sort: mockSort,
  listOfVideo: [],
  addToFavorite: mockAddToFavorite,
  removeFromFavorite: mockRemoveFromFavorite,
  deleteVideo: mockDeleteVideo
}


it('initialize Pagination with correct initial state', () => {

  const initialState = {
    showFav: false,
    startIndex: 0,
    elementsPerPage: 10,
    totalPages: 1,
    currentPage: 1
  }


  const wrapper = shallow(<Pagination {...props}/>);
  expect(wrapper.state()).toEqual(initialState);
});

it('renders properly', () => {
  const tree = TestRenderer.create(
    <Pagination {...props}/>).toJSON();
  expect(tree).toMatchSnapshot();
});

describe('methods work correctly', () => {
  const wrapper = shallow(<Pagination {...props}/>);
  const list = [
    {id: "hY7m5jjJ9mM", favorite: false},
    {id: "mRf3-JkwqfU", favorite: false},
    {id: "SB-qEYVdvXA", favorite: false}
  ];

  it('returnToList should change state showFav to false', () => {
    wrapper.setState({showFav: true});
    const instance = wrapper.instance();
    instance.returnToList();
    expect(wrapper.state('showFav')).toBe(false);
  });

  it('beforeShowDemo should call mockShowDemo and change state showFav to false', () => {
    wrapper.setState({showFav: true});
    const instance = wrapper.instance();
    instance.beforeShowDemo();
    expect(mockShowDemo.mock.calls.length).toBe(1);
    expect(wrapper.state('showFav')).toBe(false);
  });

  it('calculateTotalPagesNumber should return total number of page', () => {
    const instance = wrapper.instance();
    let result  = instance.calculateTotalPagesNumber(18, 5);
    let result2 = instance.calculateTotalPagesNumber(20, 4);
    expect(result).toBe(4);
    expect(result2).toBe(5);
  });

  it('filterList return given array if showFav is false', () => {
    const instance = wrapper.instance();
    expect(instance.filterList(list).length).toBe(3);
  });

  it('filterList return filtered array if showFav is true', () => {
    let list2 = [
      {id: "hY7m5jjJ9mM", favorite: true },
      {id: "mRf3-JkwqfU", favorite: false },
      {id: "SB-qEYVdvXA", favorite: true }
    ];
    wrapper.setState({showFav: true});
    const instance = wrapper.instance();
    expect(instance.filterList(list2).length).toBe(2);
  });

  it('setStartIndex setState properly when given argument is 1', () => {
    wrapper.setState({startIndex: 8, currentPage: 4});
    const instance = wrapper.instance();
    instance.setStartIndex(1);
    expect(wrapper.state('startIndex')).toBe(0);
    expect(wrapper.state('currentPage')).toBe(1);
  });

  it('setStartIndex setState properly when given argument is not 1', () => {
    wrapper.setState({startIndex: 12, currentPage: 4, elementsPerPage: 4 });
    const instance = wrapper.instance();
    instance.setStartIndex(3);
    expect(wrapper.state('startIndex')).toBe(8);
    expect(wrapper.state('currentPage')).toBe(3);
  });

  it('setIndexes setState properly with given number of elements per page', () => {
    wrapper.setProps({listOfVideo: list});
    const instance = wrapper.instance();
    instance.setIndexes(2);
    expect(wrapper.state('elementsPerPage')).toBe(2);
    expect(wrapper.state('startIndex')).toBe(0);
    expect(wrapper.state('currentPage')).toBe(1);
  });

  it('showFav works as expected', () => {
    let list2 = [
      {id: "hY7m5jjJ9mM", favorite: true },
      {id: "mRf3-JkwqfU", favorite: false },
      {id: "SB-qEYVdvXA", favorite: true }
    ];
    wrapper.setState({showFav: false, elementsPerPage: 4});
    wrapper.setProps({listOfVideo: list2});
    const instance = wrapper.instance();
    instance.showFavorite();
    expect(wrapper.state('showFav')).toBe(true);
    expect(wrapper.state('startIndex')).toBe(0);
    expect(wrapper.state('currentPage')).toBe(1);
    expect(wrapper.state('totalPages')).toBe(1);
  });

  it('checkForUpdate works if empty list', () => {
    wrapper.setState({totalPages: 3, currentPage: 2, startIndex: 8, elementsPerPage: 4});
    wrapper.setProps({listOfVideo:[]});
    const instance = wrapper.instance();
    instance.checkForUpdate();
    expect(wrapper.state('startIndex')).toBe(0);
    expect(wrapper.state('currentPage')).toBe(1);
    expect(wrapper.state('totalPages')).toBe(1);
    expect(wrapper.state('elementsPerPage')).toBe(10);
  });

  it('checkForUpdate works  if startIndex greater then length of list', () => {
    wrapper.setState({totalPages: 3, currentPage: 3, startIndex: 4, elementsPerPage: 2, showFav: false});
    wrapper.setProps({listOfVideo: [
      {id: "hY7m5jjJ9mM", favorite: true },
      {id: "mRf3-JkwqfU", favorite: false },
      {id: "SB-qEYVdvXA", favorite: true },
      {id: "SB-qEYVdvXB", favorite: true}
    ]});
    const instance = wrapper.instance();
    instance.checkForUpdate();
    expect(wrapper.state('startIndex')).toBe(2);
    expect(wrapper.state('currentPage')).toBe(2);
    expect(wrapper.state('totalPages')).toBe(2);
  });

  it('checkForUpdate works if new length of list is greater then it was', () => {
    wrapper.setState({totalPages: 1, currentPage: 1, startIndex: 0, elementsPerPage: 2, showFav: false});
    wrapper.setProps({listOfVideo: list});
    const instance = wrapper.instance();
    instance.checkForUpdate();
    expect(wrapper.state('startIndex')).toBe(0);
    expect(wrapper.state('currentPage')).toBe(1);
    expect(wrapper.state('totalPages')).toBe(2);
  });
});
