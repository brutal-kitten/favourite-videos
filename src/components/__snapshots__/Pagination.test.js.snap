// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<div
  className="page"
>
  <div
    className="paginationBox"
  >
    <div
      className="elementsPerPagePicker"
    >
      <span
        className="pickerText"
      >
        Elements on page
      </span>
      <select
        id="selectNumber"
        onChange={[Function]}
        tabIndex="0"
        value={10}
      >
        <option
          value="10"
        >
          10
        </option>
        <option
          value="8"
        >
          8
        </option>
        <option
          value="6"
        >
          6
        </option>
        <option
          value="4"
        >
          4
        </option>
        <option
          value="2"
        >
          2
        </option>
      </select>
    </div>
    <div
      className="pageInfo"
    >
      <div
        className="previousPage"
      >
        <button
          className="buttonPreviousPage"
          onClick={[Function]}
          tabIndex="0"
        >
          <span
            className="glyphicon glyphicon-chevron-left"
          />
        </button>
      </div>
      <div
        className="pageInput"
      >
        <input
          id="selectPage"
          max={1}
          min="1"
          onChange={[Function]}
          tabIndex="0"
          type="number"
          value={1}
        />
      </div>
      <div
        className="totalPage"
      >
        of
      </div>
      <span
        className="totalPagesNumber"
      >
        1
      </span>
      <div
        className="nextPage"
      >
        <button
          className="buttonNextPage"
          onClick={[Function]}
          tabIndex="0"
        >
          <span
            className="glyphicon glyphicon-chevron-right"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    className="list"
  >
    <h3>
      My list
    </h3>
    <div
      className="navButtons grid-container2"
    >
      <div
        className="showDemo"
      >
        <button
          className="buttonShowDemo"
          onClick={[Function]}
          tabIndex="0"
          type="button"
        >
          Show demolist
        </button>
      </div>
      <div
        className="deleteList"
      >
        <button
          className="buttonDeleteList"
          onClick={[Function]}
          tabIndex="0"
          type="button"
        >
          Delete list
        </button>
      </div>
      <div
        className="showFav"
      >
        <button
          className="buttonShowFav"
          onClick={[Function]}
          tabIndex="0"
          type="button"
        >
          Show favorite
        </button>
      </div>
      <div
        className="sort"
      >
        <span
          className="sortText"
        >
          Sort by
        </span>
        <select
          id="select"
          onChange={[Function]}
          tabIndex="0"
          value="new"
        >
          <option
            value="new"
          >
            Newest added
          </option>
          <option
            value="old"
          >
            Oldest added
          </option>
        </select>
      </div>
    </div>
    <div
      className="listContainer"
    />
  </div>
</div>
`;

exports[`renders properly 2`] = `
<div
  className="page"
>
  <div
    className="paginationBox"
  >
    <div
      className="elementsPerPagePicker"
    >
      <span
        className="pickerText"
      >
        Elements on page
      </span>
      <select
        id="selectNumber"
        onChange={[Function]}
        tabIndex="0"
        value={10}
      >
        <option
          value="10"
        >
          10
        </option>
        <option
          value="8"
        >
          8
        </option>
        <option
          value="6"
        >
          6
        </option>
        <option
          value="4"
        >
          4
        </option>
        <option
          value="2"
        >
          2
        </option>
      </select>
    </div>
    <div
      className="pageInfo"
    >
      <div
        className="previousPage"
      >
        <button
          className="buttonPreviousPage"
          onClick={[Function]}
          tabIndex="0"
        >
          <span
            className="glyphicon glyphicon-chevron-left"
          />
        </button>
      </div>
      <div
        className="pageInput"
      >
        <input
          id="selectPage"
          max={1}
          min="1"
          onChange={[Function]}
          tabIndex="0"
          type="number"
          value={1}
        />
      </div>
      <div
        className="totalPage"
      >
        of
      </div>
      <span
        className="totalPagesNumber"
      >
        1
      </span>
      <div
        className="nextPage"
      >
        <button
          className="buttonNextPage"
          onClick={[Function]}
          tabIndex="0"
        >
          <span
            className="glyphicon glyphicon-chevron-right"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    className="list"
  >
    <h3>
      My list
    </h3>
    <div
      className="navButtons grid-container2"
    >
      <div
        className="showDemo"
      >
        <button
          className="buttonShowDemo"
          onClick={[Function]}
          tabIndex="0"
          type="button"
        >
          Show demolist
        </button>
      </div>
      <div
        className="deleteList"
      >
        <button
          className="buttonDeleteList"
          onClick={[Function]}
          tabIndex="0"
          type="button"
        >
          Delete list
        </button>
      </div>
      <div
        className="showFav"
      >
        <button
          className="buttonShowFav"
          onClick={[Function]}
          tabIndex="0"
          type="button"
        >
          Show favorite
        </button>
      </div>
      <div
        className="sort"
      >
        <span
          className="sortText"
        >
          Sort by
        </span>
        <select
          id="select"
          onChange={[Function]}
          tabIndex="0"
          value="new"
        >
          <option
            value="new"
          >
            Newest added
          </option>
          <option
            value="old"
          >
            Oldest added
          </option>
        </select>
      </div>
    </div>
    <div
      className="listContainer"
    />
  </div>
</div>
`;
